lib/api/confighandler.go

```diff
@@ -0,0 +1,348 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+const configBase = "/rest/config/"
+
+type configHandler struct {
+	*http.ServeMux
+	id  protocol.DeviceID
+	cfg config.Wrapper
+	mut sync.Mutex
+}
+
+func newConfigHandler(id protocol.DeviceID, cfg config.Wrapper) http.Handler {
+	c := &configHandler{
+		ServeMux: http.NewServeMux(),
+		id:       id,
+		cfg:      cfg,
+		mut:      sync.NewMutex(),
+	}
+
+	c.HandleFunc(configBase, c.handleConfig)
+	c.HandleFunc(configBase+"insync", c.handleConfigInsync)
+	c.HandleFunc(configBase+"folders", c.handleFolders)
+	c.HandleFunc(configBase+"folders/", c.handleFolder)
+	c.HandleFunc(configBase+"devices", c.handleDevices)
+	c.HandleFunc(configBase+"devices/", c.handleDevice)
+	c.HandleFunc(configBase+"options", c.handleOptions)
+	c.HandleFunc(configBase+"ldap", c.handleLDAP)
+	c.HandleFunc(configBase+"gui", c.handleGUI)
+
+	// Legacy
+	c.HandleFunc("/rest/system/config", c.handleConfig)
+	c.HandleFunc("/rest/system/config/insync", c.handleConfigInsync)
+
+	return c
+}
+
+func (c *configHandler) handleConfig(w http.ResponseWriter, r *http.Request) {
+	if !checkMethod(w, r) {
+		return
+	}
+	if r.Method == http.MethodGet {
+		sendJSON(w, c.cfg.RawCopy())
+		return
+	}
```

Comment:
I wonder if there couldn't be some abstraction that makes this easier to read
```
type HttpHandler struct {
   Get http.Handler
   Post
   Patch
}
```

then you do 

something like
```
m.HandleFunc("devices", fromHandler(HttpHandler{
    Get: withExtractFolderId(func(foo bar) { logic })
   Post: withExtractFolderId(func(bar, nuff) { logic })
})

func withExtractFolderIf(delegate func(folder, w, r)) func(w, r) {
    return func(w, r) {
         fid := extractFolderId(r)
         delegate(fid, w, r)
    }
))
``` 
any function pointer that's not set returns method not allowed.

Now there is a bunch of `if method == get: sentJson(); return` 

Or I mean, there are frameworks for route and method handling, just use a framework instead of reinventing the wheel, of stripping paths to extract folder ids etc.

Reply:
I didn't fully get to something just like you propose, but I tried to go into that direction - is readability any better? And used a framework, which is need indeed, I usually am just vary to add a dependency for a tiny usage like this.